src/linquantarrays.jl

```diff
@@ -1,105 +1,187 @@
+
+    A = similar(Q, U)
 
     @inbounds for i in eachindex(A)
-        # convert Q[i]::UInt to Float64 via *
+        # convert Q[i]::Integer to Float64 via *
         # then to T through =
-        A[i] = Qmin + Q[i]*Δ
+        A[i] = Qmin + (Q[i] - Tmin)*Δ
     end
 
     return A
 end
 
-# define default conversions for 8, 16, 24 and 32 bit
-Base.Array{T}(Q::LinQuantArray{UInt8,N}) where {T,N} = Array{T}(8,Q)
-Base.Array{T}(Q::LinQuantArray{UInt16,N}) where {T,N} = Array{T}(16,Q)
-Base.Array{T}(Q::LinQuantArray{UInt24,N}) where {T,N} = Array{T}(24,Q)
-Base.Array{T}(Q::LinQuantArray{UInt32,N}) where {T,N} = Array{T}(32,Q)
-
-Base.Array(Q::LinQuantArray{UInt8,N}) where N = Array{Float32}(8,Q)
-Base.Array(Q::LinQuantArray{UInt16,N}) where N = Array{Float32}(16,Q)
-Base.Array(Q::LinQuantArray{UInt24,N}) where N = Array{Float32}(24,Q)
-Base.Array(Q::LinQuantArray{UInt32,N}) where N = Array{Float64}(32,Q)
-
-# one quantization per layer
-"""Linear quantization independently for every element along dimension
-dim in array A. Returns a Vector{LinQuantArray}."""
-function LinQuantArray(::Type{TUInt},A::AbstractArray{T,N},dim::Int) where {TUInt,T,N}
+# default conversions for unsigned 8, 16, 24 and 32 bit
+Base.Array(Q::LinQuantArray{UInt8,N}) where {N} = Array{Float32}(Q)
+Base.Array(Q::LinQuantArray{UInt16,N}) where {N} = Array{Float32}(Q)
+Base.Array(Q::LinQuantArray{UInt24,N}) where {N} = Array{Float32}(Q)
+Base.Array(Q::LinQuantArray{UInt32,N}) where {N} = Array{Float64}(Q)
+
+# default conversions for signed 8, 16, 24 and 32 bit
+Base.Array(Q::LinQuantArray{Int8,N}) where N = Array{Float32}(Q)
+Base.Array(Q::LinQuantArray{Int16,N}) where N = Array{Float32}(Q)
+Base.Array(Q::LinQuantArray{Int24,N}) where N = Array{Float32}(Q)
+Base.Array(Q::LinQuantArray{Int32,N}) where N = Array{Float64}(Q)
+
+
+"""
+    LinQuantArray(TInteger, A, dim; extrema=nothing)
+
+Linear quantization independently for every element along dimension
+dim in array A.
+
+# Arguments
+- `TInteger`: the type of the quantised array
+- `A`: the array to quantise
+- `dim`: the dimension along which to quantise
+- `extrema`: the minimum and maximum of the range, defaults to `nothing`.
+
+# Returns
+- a Vector{LinQuantArray} with the quantised array and the minimum and maximum of the original range.
+"""
+function LinQuantArray(
+    ::Type{TInteger},
+    A::AbstractArray{T,N},
+    dim::Integer;
+    extrema::Option{Tuple} = nothing
+) where {TInteger,T,N}
     @assert dim <= N   "Can't quantize a $N-dimensional array in dim=$dim"
     n = size(A)[dim]
-    L = Vector{LinQuantArray}(undef,n)
+    L = Vector{LinQuantArray}(undef, n)
     t = [if j == dim 1 else Colon() end for j in 1:N]
     for i in 1:n
         t[dim] = i
-        L[i] = LinQuantization(TUInt,A[t...])    
+        L[i] = LinQuantization(TInteger,A[t...]; extrema=extrema)    
     end
     return L
 end
 
-# for 8,16,24 and 32 bit
+function LinQuantArray{U}(
+    A::AbstractArray{T,N},
+    dim::Int; 
```

Comment:
I guess technically one would want for consistency to use a `dims` keyword, similar to how sorting along a dimension works
```julia
julia> sort(randn(5,5), dims=1)
5×5 Matrix{Float64}:
 -1.76503   -1.1454    -0.666234   -1.60678    -1.93496
 -1.09311   -0.555642  -0.374038   -1.01704    -1.10898
 -0.970622   0.138932  -0.353606    0.0905675  -0.414778
  0.505033   0.259424  -0.0457179   0.386296    0.0198439
  1.25263    1.52726    1.32485     1.93694     2.14597
```

but I'm happy either way for now, just though that `LinQuantArray(A, 2)` isn't as verbose as `LinQuantArray(A, dims=2)`

Reply:
But it would be a breaking change, wouldn't it?

Reply:
it would be breaking once we agree on one way and decide to do the other, yes.

Reply:
I find it better to use `dims` and be consistent with other Julia functions.